integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/minikube/Minikube.scala

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s.integrationtest.backend.minikube
+
+import java.nio.file.Paths
+
+import io.fabric8.kubernetes.client.{ConfigBuilder, DefaultKubernetesClient}
+
+import org.apache.commons.lang3.SystemUtils
+import org.apache.spark.deploy.k8s.integrationtest.{Logging, ProcessUtils}
+
+// TODO support windows
+private[spark] object Minikube extends Logging {
+  private val MINIKUBE_EXECUTABLE_DEST = if (SystemUtils.IS_OS_MAC_OSX) {
+    Paths.get("target", "minikube-bin", "darwin-amd64", "minikube").toFile
+  } else if (SystemUtils.IS_OS_WINDOWS) {
+    throw new IllegalStateException("Executing Minikube based integration tests not yet " +
+      " available on Windows.")
+  } else {
+    Paths.get("target", "minikube-bin", "linux-amd64", "minikube").toFile
+  }
+
+  private val EXPECTED_DOWNLOADED_MINIKUBE_MESSAGE = "Minikube is not downloaded, expected at " +
+    s"${MINIKUBE_EXECUTABLE_DEST.getAbsolutePath}"
+
+  private val MINIKUBE_STARTUP_TIMEOUT_SECONDS = 60
+
+  def startMinikube(): Unit = synchronized {
```

Comment:
Why are this and the following methods `synchronized `? `executeMinikube ` cannot be called concurrently?

Reply:
@mccheah would know better. I guess we want to prevent `deleteMinikube` from destroying the minikube vm while other methods try to use the vm. Maybe such a race condition can corrupt the vm or some vm provisioning tools like `VirtualBox`?

Does this explanation make sense? If yes, I think we can leave a NOTE.

Reply:
It makes sense to me.